### ISSUE TITLE
CUDA OOM Issue When Using Approx Tanh with softcapping score mod

### ISSUE THREAD

**Issue by kebijuelun** [pytorch-labs/attention-gym#43]
I've encountered a CUDA out-of-memory (OOM) issue in the following scenario while working with the code provided below:

- reproduce code 

```python
import torch
import random
from functools import partial
from typing import Any

from torch.nn.attention.flex_attention import create_block_mask
from torch.nn.attention.flex_attention import BlockMask, _mask_mod_signature, _score_mod_signature, flex_attention
from torch.nn.attention.flex_attention import _mask_mod_signature

from functools import partial

import torch
from torch._inductor.lowering import make_pointwise, register_lowering
from torch._inductor.virtualized import ops

flex_attention = torch.compile(flex_attention, dynamic=False)


@torch.library.custom_op('approx::tanh', mutates_args=())
def _tanh_approx(inp: torch.Tensor) -> torch.Tensor:
    return torch.tanh(inp)


@_tanh_approx.register_fake
def _(inp: torch.Tensor) -> torch.Tensor:
    return torch.tanh(inp)


def _tanh_approx_lowering(inp):
    fn = partial(ops.inline_asm_elementwise, asm='tanh.approx.f32 $0, $1;')
    return make_pointwise(fn)(inp)


register_lowering(torch.ops.approx.tanh)(_tanh_approx_lowering)


class TanhApprox(torch.autograd.Function):
    """Custom autograd function for tanh approximation.

    Code taken from: https://github.com/pytorch-labs/attention-gym/blob/75867424a1d4391bff49527029d3612a09dd67e2/attn_gym/mods/softcapping.py
    """

    @staticmethod
    def forward(x):
        return torch.ops.approx.tanh(x)

    @staticmethod
    def setup_context(ctx, inputs, output):
        (x, ) = inputs
        result = output
        ctx.save_for_backward(result)

    @staticmethod
    def backward(ctx, grad_output):
        (result, ) = ctx.saved_tensors
        return grad_output * (1 - result * result)

    @staticmethod
    def vmap(info, in_dims, x):
        return torch.tanh(x), 0


def _offsets_to_doc_ids_tensor(offsets):
    device = offsets.device
    counts = offsets[1:] - offsets[:-1]
    return torch.repeat_interleave(torch.arange(len(counts), device=device, dtype=torch.int32), counts)


def generate_doc_mask_mod_full(offsets: torch.Tensor) -> _mask_mod_signature:
    """Generates full attention mask mods that apply to inputs to flex attention in the sequence stacked
    format.

    Args:
        offsets: This tensor should be of shape(num_documents + 1)
            this should contain the cumulative counts of document tokens.
            e.g. if you have 3 documents of length 2, 4, 3 then
            offsets = [0, 2, 6, 9]
    """
    document_id = _offsets_to_doc_ids_tensor(offsets)

    def doc_mask_mod(b, h, q_idx, kv_idx):
        same_doc = document_id[q_idx] == document_id[kv_idx]
        return same_doc

    return doc_mask_mod


_tanh_approx = TanhApprox.apply


def generate_tanh_softcap(soft_cap: int, approx: bool = False) -> _score_mod_signature:
    """Returns an tanh bias score_mod given the number of heads H.

    Args:
        soft_cap: The soft cap value to use for normalizing logits
        approx: Whether to use the `tanh.approx.` ptx instruction

    Returns:
        tanh_softcap: score_mod
    """
    tanh = _tanh_approx if approx else torch.tanh

    def tanh_softcap(score, b, h, q_idx, kv_idx):
        return soft_cap * tanh(score / soft_cap)

    prefix = 'tanh_softcap_approx' if approx else 'tanh_softcap'
    tanh_softcap.__name__ = f'{prefix}_{soft_cap}'

    return tanh_softcap


def flex_attn_varlen(q: torch.Tensor,
                     k: torch.Tensor,
                     v: torch.Tensor,
                     block_mask: BlockMask,
                     score_mod: _score_mod_signature | None = None,
                     logits_cap: float | None = None,
                     kernel_options: dict[str, Any] | None = None):
    """
    Args:
        q: Query tensor of shape (num_heads, seqlen, head_dim).
        k: Key tensor of shape (num_heads, seqlen, head_dim).
        v: Value tensor of shape (num_heads, seqlen, head_dim).
        score_mod: Function to modify attention scores. By default no score_mod is applied.
        block_mask: BlockMask object that controls the blocksparsity pattern of the attention.
        logits_cap: Use the logits cap or not.
        kernel_options: Options to pass into the Triton kernels.
    """
    q, k, v = q.unsqueeze(0), k.unsqueeze(0), v.unsqueeze(0)
    assert isinstance(block_mask, BlockMask), block_mask

    if logits_cap is not None and logits_cap > 1e-6:
        # Generate tanh-based soft cap for logits
        # TODO: score_mod merge not yet supported
        assert score_mod is None, 'score_mod must be None when using logits_cap'
        score_mod = generate_tanh_softcap(logits_cap, approx=True)
        # score_mod = generate_tanh_softcap(logits_cap, approx=False)

    o = flex_attention(q, k, v, score_mod=score_mod, block_mask=block_mask, kernel_options=kernel_options)
    o = o.squeeze(0)
    return o.to(q.dtype)


def generate_qkv_from_shapes(seqlen, dtype, flex_order=False):
    random.seed(0)
    torch.manual_seed(0)
    torch.cuda.manual_seed(0)
    device = torch.cuda.current_device()


    # gen qkv, num heads and head dim are settled
    if flex_order:
        q = torch.randn((8, seqlen, 128), dtype=dtype, device=device, requires_grad=True)
        k = torch.randn((8, seqlen, 128), dtype=dtype, device=device, requires_grad=True)
        v = torch.randn((8, seqlen, 128), dtype=dtype, device=device, requires_grad=True)
    else:
        q = torch.randn((seqlen, 8, 128), dtype=dtype, device=device, requires_grad=True)
        k = torch.randn((seqlen, 8, 128), dtype=dtype, device=device, requires_grad=True)
        v = torch.randn((seqlen, 8, 128), dtype=dtype, device=device, requires_grad=True)

    return q, k, v

lengths = [256, 2048, 1024, 6144, 2048, 6144, 1792, 512, 6656, 86400]
seqlens = torch.tensor([0] + lengths, dtype=torch.int32, device='cuda')
cu_seqlens = seqlens.cumsum(dim=0, dtype=torch.int32)

q, k, v = generate_qkv_from_shapes(sum(lengths), dtype=torch.bfloat16, flex_order=True)
full_mask_mod = generate_doc_mask_mod_full(cu_seqlens)
S = cu_seqlens[-1].item()
full_block_mask = create_block_mask(full_mask_mod, 1, 1, S, S, device=q.device, _compile=True)
with torch.set_grad_enabled(False):
# with torch.set_grad_enabled(True):
    out = flex_attn_varlen(q, k, v, full_block_mask, logits_cap=80)
```

However, if I make either of these changes, the OOM issue does not occur:
- Change torch.set_grad_enabled(False) to torch.set_grad_enabled(True)
- Change score_mod = generate_tanh_softcap(logits_cap, approx=True) to score_mod = generate_tanh_softcap(logits_cap, approx=False)

Key Details:
- Approximation seems to be tied to the issue, specifically when combined with disabling gradients.
- CUDA OOM doesn't occur when using the standard torch.tanh() instead of the approximate version.
- Potential interaction between the custom approximation and PyTorch's memory management in inference mode (i.e., when gradients are disabled).

Environment:
- torch                       2.5.0.dev20240907+cu121

Questions:

- What might be causing the CUDA OOM specifically when using approx=True and torch.set_grad_enabled(False)?
- Is there a recommended approach to avoid the OOM issue without disabling gradient tracking or using the non-approximate tanh?

**Comment by drisspg** [comment]
If you have Pytorch checked out, would you mind running `python torch/utils/collect_env.py`

**Comment by kebijuelun** [comment]
> python torch/utils/collect_env.py

sorry for the late reply, my env collection:

```
Collecting environment information...
PyTorch version: 2.5.0.dev20240907+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.3 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: version 3.22.1
Libc version: glibc-2.35

Python version: 3.10.14 (main, May  6 2024, 19:42:50) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.10.135-6-velinux1u1-amd64-x86_64-with-glibc2.35
Is CUDA available: False
CUDA runtime version: 12.1.105
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration:
Nvidia driver version: Could not collect
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.9.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
...

Versions of relevant libraries:
[pip3] flake8==6.1.0
[pip3] intel-extension-for-pytorch==2.2.0
[pip3] msgpack-numpy==0.4.8
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.26.4
[pip3] open_clip_torch==2.26.1
[pip3] pytorch-triton==3.0.0+757b6a61e7
[pip3] pytorchvideo==0.1.5
[pip3] torch==2.5.0.dev20240907+cu121
[pip3] torchaudio==2.5.0.dev20240907+cu121
[pip3] torchvision==0.20.0.dev20240907+cu121
[conda] intel-extension-for-pytorch 2.2.0                    pypi_0    pypi
[conda] msgpack-numpy             0.4.8                    pypi_0    pypi
[conda] numpy                     1.26.4                   pypi_0    pypi
[conda] open-clip-torch           2.26.1                   pypi_0    pypi
[conda] pytorch-triton            3.0.0+757b6a61e7          pypi_0    pypi
[conda] torch                     2.5.0.dev20240907+cu121          pypi_0    pypi
[conda] torchaudio                2.5.0.dev20240907+cu121          pypi_0    pypi
[conda] torchvision               0.20.0.dev20240907+cu121          pypi_0    pypi
```

**Comment by drisspg** [comment]
Sorry, what GPU are you using and how much global memory does it have?

**Comment by kebijuelun** [comment]
> Sorry, what GPU are you using and how much global memory does it have?

Sorry for the late reply. My GPU is an A100, featuring 80GB of global memory.

**Comment by drisspg** [comment]
SO the problem is that there was a silent graph_break

can you set fullgraph=True and try. [137809](https://github.com/pytorch/pytorch/pull/137809)
Should have fixed the issue, and I am unable to repro on nightly

**Comment by drisspg** [comment]
clsoing as unable to repro
